File: qs_java/Quicksort.java
import java.util.List;
import java.util.ArrayList;

public class Quicksort {
    public static void quicksort(List<Integer> unsorted_list) {
        int size = unsorted_list.size();

        if (size > 1) {
            int pivot = unsorted_list.get(0);
            int L_idx = 0;
            int R_idx = size - 1; 

            while (L_idx < R_idx) {
                while (unsorted_list.get(L_idx) < pivot && L_idx < size) {
                    L_idx++;
                }
                while (unsorted_list.get(R_idx) > pivot && R_idx > 0) {
                    R_idx--;
                }
                int temp = unsorted_list.get(L_idx);
                unsorted_list.set(L_idx, unsorted_list.get(R_idx));
                unsorted_list.set(R_idx, temp);
            }
            quicksort(unsorted_list.subList(0, L_idx + 1));
            quicksort(unsorted_list.subList(L_idx + 1, size));
        }
    }

    public static void main(String[] args) {
        List<Integer> test_list = new ArrayList<Integer>();
        test_list.add(4);
        test_list.add(8);
        test_list.add(1);
        test_list.add(6);
        test_list.add(3);
        test_list.add(7);
        test_list.add(2);
        test_list.add(5);
        System.out.println("Unsorted list: " + test_list);
        quicksort(test_list);
        System.out.println("\nSorted list:   " + test_list);
    }
}

File: qs_c/quicksort.c
#include <stdio.h>

int quicksort(int *unsorted_list, int size) {

    /* Base case: nothing left to partition. */
    if (size < 2) {
        return 0;
    }
    else {
        /* First, select a pivot. */
        int pivot = unsorted_list[0];
        /* Set the left pointer to the leftmost index, which will
           always be 0. */
        int L_idx = 0;
        /* Set the right pointer to the rightmost index, which will
           always be the size of the array - 1. */
        int R_idx = size - 1;
        /* Swap elements until the L and R idx collide. */
        while (L_idx < R_idx) {
            /* Increment L_idx until its value is greater than pivot. */
            while (unsorted_list[L_idx] < pivot && L_idx < size) {
                L_idx++;
            }
            /* Decrement R_idx until its value is less than pivot. */
            while (unsorted_list[R_idx] > pivot && R_idx > 0) {
                R_idx--; 
            }
            /* Swap the left and right elements. */
            int temp = unsorted_list[L_idx];
            unsorted_list[L_idx] = unsorted_list[R_idx];
            unsorted_list[R_idx] = temp;
        }
        /* Sort the left partition. */
        quicksort(unsorted_list, L_idx);
        /* Sort the right partition. */
        quicksort(&unsorted_list[L_idx + 1], size - L_idx - 1);
        
        return 0;
    }
} 

int print_list(int *list, int size) {
    int i = 0;

    for (i = 0; i < size; i++) {
        printf("%d ", list[i]);
    }

    printf("\n");

    return 0;
}

int main(void) {
    int list_size = 8;
    int test_list[] = {4, 8, 1, 6, 3, 7, 2, 5};
    printf("Unsorted list: ");
    print_list(test_list, list_size);
    quicksort(test_list, list_size);
    printf("\nSorted list:   ");
    print_list(test_list, list_size);

    return 0;
}
